smolyay/samples.py

```diff
@@ -508,3 +508,620 @@ def _create(self):
+
+    def _get_random_points(self):
+        """Generate a set of random points
+
+        Generates a set of quasi-monte carlo Sobol Sequence points.
+        """
+        lower_bounds = [bound[0] for bound in self.domain]
+        upper_bounds = [bound[1] for bound in self.domain]
+        p_gen = scipy.stats.qmc.Sobol(
+            self.num_dimensions,
+            scramble=self.scramble,
+            bits=self.bits,
+            optimization=self.optimization,
+            seed=self.seed,
+        ).random(n=self.number_points)
+        return scipy.stats.qmc.scale(p_gen, lower_bounds, upper_bounds)
+
+
+class PointSetProduct(MultidimensionalPointSet):
+    """Generate a grid using combinations of a unidimensional point set
+
+    Parameters
+    ----------
+    point_sets : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def __init__(self, point_sets):
+        super().__init__()
+        self._point_sets = point_sets
+        self._indexes = []
+
+    @property
+    def domain(self):
+        """numpy.ndarray: domain of the point set."""
+        return numpy.array([up.domain for up in self.point_sets])
+
+    @domain.setter
+    def domain(self, value):
+        domain = numpy.sort(numpy.array(value, ndmin=2), axis=1)
+        if domain.ndim != 2 or domain.shape[1] != 2:
+            raise TypeError("Domain must have size (num_dimensions, 2)")
+        if any(domain[:, 0] >= domain[:, 1]):
+            raise ValueError("Lower bound must be less than upper bound")
+        if len(self.point_sets) != domain.shape[0]:
+            raise IndexError("Domain does not match number of point sets")
+        if not numpy.array_equal(self._domain, domain):
+            self._domain = domain
+            for ps,d in zip(self.point_sets,domain):
+                ps.domain = d
+            self._valid_cache = False
+
+    @property
+    def point_sets(self):
+        """:class:UnidimensionalPointSet: set of unique points"""
+        return self._point_sets
+
+    @property
+    def indexes(self):
+        """list: Grid points' indexes."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._indexes
+
+
+class TensorProductPointSet(PointSetProduct):
+    """Generate a grid using all combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        self._points = numpy.array(list(itertools.product(*self._point_sets)))
```

Comment:
This is going to cause a double memory allocation since it makes a list then an array. I think something like this would be good:
```suggestion
        num_points = numpy.prod([len(p) for p in self._point_sets])
        self._points = numpy.array((num_points, self.num_dimensions), dtype=float)
        for i, point in enumerate(itertools.product(*self._point_sets)):
            self._points[i] = point
```

Reply:
What about doing something like this instead?
It worked when I tested it, and the runtime doesn't grow as fast.
`self._points = numpy.array(numpy.meshgrid(*self._point_sets)).T.reshape(-1, self.num_dimensions)`

Reply:
Using self.num_dimensions by necessity creates a dependency on the self.domain property, which for PointSetProduct creates a dependency on the elements in self._point_sets having a domain property.

Overridding the getter for num_dimensions in the PointSetProduct to `return len(self.point_sets)` removes that dependency. You could put in a list of numpy arrays as the point_sets and the tensor would be able to generate grid points just fine. The UnidimensionalPointSets can be used like lists, and while that's not always the case in reverse, it would be very trivial to make the tensor capable of using pure lists in place of unidimensional point sets.

Reply:
You could add `num_dimensions = len(self._point_sets)` in my code above instead of `self.num_dimensions`, rather than need to touch aany other code. Unless there is a drastic time bottleneck, I think you are better off using the lower memory version I suggested. Your array and meshgrid will probably trigger at least 1 extra allocation, and this is bad for large tensor products.

Reply:
What about 

`self._points = numpy.stack(numpy.meshgrid(*self._point_sets), -1).reshape(
            -1, len(self._point_sets))`

It outputs the points in a different order but it ended up being slightly faster in the higher dimensions than the first meshgrid idea.

Reply:
Is there a reason you can’t use the code as I wrote it? It uses an iterator so the array does not get generated twice. I am pretty sure yours will due to the combination of meshgrid with other array operations. Memory considerations are very important when arrays can get large, which is the case for the tensor product.

Reply:
Simply that using the iterator has a worse runtime complexity than either of the meshgrid strategies. If there was some strategy and combination of operations that allowed the best of both worlds, the meshgrid strategies runtime complexity without significant memory allocations, then I would want that one. I am just trying to see if I can find a strategy that doesn't make extra copies of the array.

As it stands, we can use the iterator strategy.

Reply:
Do you have a timing that shows how much slower the iterator is? In general, you shouldn’t optimize until you profile and find something is a bottleneck. In this case, I don’t really care if this bit of code is a little slower because it will typically be done only once and so will probably be a negligible cost most of the time. I do care, though, if the calculation gets killed that one time it runs because Python ran out of memory.

Just FYI: complexity and time aren’t the same thing. The complexity of both approaches is O(N), I believe, but one is just slower because it isn’t vectorized by NumPy.

Reply:
I do. I should be able to show it tomorrow.

Reply:
I did 2 test runs with all the different methods (yours with itertools, using meshgrid and transpose, and using meshgrid and stack).
![image](https://github.com/biomodsquad/smolyak-surrogate-modeling/assets/107584654/31268249-7187-486e-983c-675cdafb5f9c)

This was the runtimes for all three methods making a 2D grid, a 3D grid, and a 4D grid. As the number of unique values used to make the grid increases, the itertools methods takes much longer and the time difference gets more drastic as the dimensions increase. I made the y axis logarithmic so the runtimes making 2D grids is visible.
![image](https://github.com/biomodsquad/smolyak-surrogate-modeling/assets/107584654/e8de8b52-34c0-4e65-9b9b-620fec9034c9)

This is a different run that includes 5D and 6D grids. The intertools method making an n dimensional grid tends to have a runtime comparable to making a n+1 dimensional grid using the 2 meshgrid methods.

In retrospect, the time differences aren't that big a deal since we're going to be mostly working on a small number of unique points, (the largest 6D grid uses only 5 unique points), but it's a time save that would be nice to have if possible.

Reply:
Given that these differences probably aren't significant to the overall problem (solving the coefficients of a 15^6 grid likely takes way longer than 1 s), we should use the `itertools` version for readability and to reduce memory usage.

Moving forward, I recommend that you don't worry about running microbenchmarks like this for now. It will be faster to get a working prototype (coded however we think is best), then profile actual use cases to see where performance optimization is needed.

You may find it interesting to read about [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law).